diff --git a/src/Makefile b/src/Makefile
index 1f1987fbe..d0114b3b8 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -10,6 +10,8 @@ PRODUCTION_VERSION      := v7.1.2
 ENABLE_CUBIN            ?= 1
 ENABLE_BRAIN            ?= 1
 ENABLE_UNRAR            ?= 1
+ENABLE_LTO              ?= 0
+USE_MOLD                ?= 0
 USE_SYSTEM_LZMA         ?= 0
 USE_SYSTEM_ZLIB         ?= 0
 USE_SYSTEM_OPENCL       ?= 0
@@ -201,12 +203,40 @@ COMPTIME                := $(shell date +%s)
 ## General compiler and linker options
 ##
 
+CC_NATIVE_CLANG         := 0
+
+ifneq (,$(findstring clang, $(CC)))
+CC_NATIVE_CLANG         := 1
+endif
+
+CFLAGS                  := $(CFLAGS)
 LFLAGS                  := $(LDFLAGS)
 
+ifeq ($(DEBUG),1)
+CFLAGS                  += -fstack-usage
+endif
+
+ifeq ($(ENABLE_LTO),1)
+CFLAGS                  += -flto=auto
+ifeq ($(CC_NATIVE_CLANG),1)
+CFLAGS                  += -Wno-unknown-warning-option
+endif
+LFLAGS                  += -flto=auto -Wno-lto-type-mismatch
+endif
+
+ifeq ($(BUILD_MODE),native)
+ifneq ($(UNAME),Darwin)
+ifeq ($(USE_MOLD),1)
+LFLAGS                  += -fuse-ld=mold
+endif
+endif
+endif
+
 ifneq ($(UNAME),Darwin)
 ifeq ($(MAINTAINER_MODE),0)
 ifeq ($(BUILD_MODE),native)
 CFLAGS                  += -march=native
+CFLAGS                  += -mtune=native
 endif
 endif
 endif
@@ -218,7 +248,7 @@ CFLAGS                  += -Wextra
 endif
 
 ## because LZMA SDK
-ifneq (,$(findstring clang, $(CC)))
+ifeq ($(CC_NATIVE_CLANG),1)
 CFLAGS                  += -Wno-typedef-redefinition
 else
 ifeq ($(CC),cc)
@@ -228,7 +258,7 @@ endif
 
 ifeq ($(USE_SYSTEM_LZMA),0)
 CFLAGS_LZMA             += -D_7ZIP_ST
-ifeq (,$(findstring clang, $(CC)))
+ifeq ($(CC_NATIVE_CLANG),0)
 CFLAGS_LZMA             += -Wno-misleading-indentation
 endif
 endif
@@ -240,7 +270,7 @@ CFLAGS_ZLIB             += -Wno-implicit-function-declaration
 CFLAGS_ZLIB             += -Wno-unused-parameter
 CFLAGS_ZLIB             += -DIOAPI_NO_64
 CFLAGS_ZLIB             += -DUNZ_BUFSIZE=262144
-ifneq (,$(findstring clang, $(CC)))
+ifeq ($(CC_NATIVE_CLANG),1)
 CFLAGS_ZLIB             += -Wno-deprecated-non-prototype -Wno-unknown-warning-option
 endif
 endif
@@ -284,8 +314,9 @@ endif
 endif
 
 ifeq ($(DEBUG),0)
-CFLAGS                  += -O2
+CFLAGS                  += -O2 -fomit-frame-pointer
 ifneq ($(UNAME),Darwin)
+CFLAGS                  += -fno-plt
 LFLAGS                  += -s
 endif
 else
@@ -354,9 +385,11 @@ CFLAGS                  += -I$(DEPS_UNRAR_PATH)
 ifeq ($(UNAME),OpenBSD)
 LFLAGS                  += -lc++
 else
+ifeq ($(UNAME),Darwin)
 LFLAGS                  += -lstdc++
 endif
 endif
+endif
 
 ##
 ## Native compilation target
@@ -552,23 +585,31 @@ clean:
 	$(RM) -rf modules/*.dSYM
 	$(RM) -rf bridges/*.dSYM
 	$(RM) -rf feeds/*.dSYM
-	$(RM) -f modules/*.dll
-	$(RM) -f bridges/*.dll
+	$(RM) -rf *.dSYM
 	$(RM) -f bridges/subs/*.dll
-	$(RM) -f feeds/*.dll
-	$(RM) -f modules/*.so
-	$(RM) -f bridges/*.so
 	$(RM) -f bridges/subs/*.so
+	$(RM) -f bridges/subs/*.su
+	$(RM) -f bridges/*.dll
+	$(RM) -f bridges/*.so
+	$(RM) -f bridges/*.su
+	$(RM) -f feeds/*.dll
 	$(RM) -f feeds/*.so
+	$(RM) -f feeds/*.su
+	$(RM) -f modules/*.dll
+	$(RM) -f modules/*.so
+	$(RM) -f modules/*.su
 	$(RM) -f obj/*/*/*.o
+	$(RM) -f obj/*/*/*.su
 	$(RM) -f obj/*/*.o
+	$(RM) -f obj/*/*.su
 	$(RM) -f obj/*.o
+	$(RM) -f obj/*.su
 	$(RM) -f obj/*.a
-	$(RM) -rf *.dSYM
 	$(RM) -f *.dylib
 	$(RM) -f *.bin *.exe
 	$(RM) -f *.pid
 	$(RM) -f *.log
+	$(RM) -f *.su
 	$(RM) -f core
 	$(RM) -rf *.induct
 	$(RM) -rf *.outfiles
@@ -914,27 +955,95 @@ WIN_PYTHON              := /opt/win-python
 ## Compiler options
 ##
 
+CC_LINUX_CLANG          := 0
+CC_WIN_CLANG            := 0
+
+ifneq (,$(findstring clang, $(CC_LINUX)))
+CC_LINUX_CLANG          := 1
+endif
+
+ifneq (,$(findstring clang, $(CC_WIN)))
+CC_WIN_CLANG            := 1
+endif
+
 CFLAGS_CROSS_LINUX      := $(CFLAGS)
 CFLAGS_CROSS_LINUX      += -fPIC
 CFLAGS_CROSS_LINUX      += -DWITH_HWMON
+
+ifeq ($(CC_LINUX_CLANG),1)
+CFLAGS_CROSS_LINUX      += -Wno-typedef-redefinition
+CFLAGS_CROSS_LINUX      += -Wno-unknown-warning-option
+endif
+
 CFLAGS_CROSS_WIN        := $(CFLAGS)
 CFLAGS_CROSS_WIN        += -fPIC
 CFLAGS_CROSS_WIN        += -I$(WIN_ICONV)/include/
 CFLAGS_CROSS_WIN        += -DWITH_HWMON
 
+ifeq ($(CC_WIN_CLANG),1)
+CFLAGS_CROSS_WIN        := $(filter-out -fno-plt,$(CFLAGS_CROSS_WIN))
+CFLAGS_CROSS_WIN        += --target=x86_64-w64-mingw32
+CFLAGS_CROSS_WIN        += -D_CLANG_DISABLE_CRT_DEPRECATION_WARNINGS
+CFLAGS_CROSS_WIN        += -D__INTRINSIC_DEFINED___cpuidex=1
+CFLAGS_CROSS_WIN        += -D__INTRINSIC_DEFINED__xgetbv=1
+CFLAGS_CROSS_WIN        += -Wno-typedef-redefinition
+CFLAGS_CROSS_WIN        += -Wno-ignored-attributes
+CFLAGS_CROSS_WIN        += -Wno-unknown-warning-option
+endif
+
 LFLAGS_CROSS_LINUX      := $(LFLAGS)
 LFLAGS_CROSS_LINUX      += -lpthread
 LFLAGS_CROSS_LINUX      += -ldl
 LFLAGS_CROSS_LINUX      += -lm
+LFLAGS_CROSS_LINUX      += -static-libgcc
+
+ifeq ($(CC_LINUX_CLANG),1)
+LFLAGS_CROSS_LINUX      += -stdlib=libc++
+else
+LFLAGS_CROSS_LINUX      += -static-libstdc++
+endif
+
+ifeq ($(USE_MOLD),1)
+LFLAGS_CROSS_LINUX      += -fuse-ld=mold
+else
+ifeq ($(ENABLE_LTO),1)
+ifeq ($(CC_LINUX_CLANG),1)
+LFLAGS_CROSS_LINUX      += -fuse-ld=lld
+else
+LFLAGS_CROSS_LINUX      += -fuse-ld=gold
+endif
+endif
+endif
+
 LFLAGS_CROSS_WIN        := $(LFLAGS)
+ifneq ($(UNAME),Darwin)
+LFLAGS_CROSS_WIN        += -lstdc++
+endif
+
+ifeq ($(CC_WIN_CLANG),0)
 LFLAGS_CROSS_WIN        += -Wl,--dynamicbase
 LFLAGS_CROSS_WIN        += -Wl,--nxcompat
+endif
 LFLAGS_CROSS_WIN        += -lpsapi
 LFLAGS_CROSS_WIN        += -lws2_32
 LFLAGS_CROSS_WIN        += -lpowrprof
-LFLAGS_CROSS_WIN        += -static -static-libgcc -static-libstdc++
+
+ifeq ($(CC_WIN_CLANG),0)
+LFLAGS_CROSS_WIN        += -static
+endif
+LFLAGS_CROSS_WIN        += -static-libgcc
+ifeq ($(CC_WIN_CLANG),0)
+LFLAGS_CROSS_WIN        += -static-libstdc++
+endif
 LFLAGS_CROSS_WIN        += -lole32 -loleaut32 -lwbemuuid
 
+ifeq ($(ENABLE_LTO),1)
+ifeq ($(CC_WIN_CLANG),1)
+LFLAGS_CROSS_WIN        += -fuse-ld=lld
+LFLAGS_CROSS_WIN        += -Wl,--stack,0x200000
+endif
+endif
+
 CFLAGS_LZMA_WIN         := $(CFLAGS_LZMA)
 CFLAGS_UNRAR_WIN        := $(CFLAGS_UNRAR)
 
