diff --git a/src/shared.c b/src/shared.c
index 4104628e5..b850b1f3b 100644
--- a/src/shared.c
+++ b/src/shared.c
@@ -1520,6 +1520,10 @@ int get_current_arch ()
 {
   #if defined (_WIN)
 
+  #if !defined(PROCESSOR_ARCHITECTURE_ARM64)
+  #define PROCESSOR_ARCHITECTURE_ARM64 12
+  #endif
+
   SYSTEM_INFO sysinfo;
 
   GetNativeSystemInfo (&sysinfo);
diff --git a/src/terminal.c b/src/terminal.c
index 541fb94f3..f02007712 100644
--- a/src/terminal.c
+++ b/src/terminal.c
@@ -1259,6 +1259,11 @@ void backend_info (hashcat_ctx_t *hashcat_ctx)
     }
 
     #if defined (_WIN)
+
+    #if !defined(PROCESSOR_ARCHITECTURE_ARM64)
+    #define PROCESSOR_ARCHITECTURE_ARM64 12
+    #endif
+
     // Get Windows system information
     SYSTEM_INFO sysinfo;
     OSVERSIONINFO osvi;
